HIP/hip-991.md

```diff
@@ -44,182 +44,310 @@ This document outlines the development of a fixed fee system for the submission
+* As a **Regular User**, I want to know the exact cost of interacting with a topic.
 
-```
-message FixedFee {
-  // The number of units to assess as a fee
-  int64 amount = 1;
-  // The denomination of the fee; taken as hbar if left unset
-  TokenID denominating_token_id = 2;
-}
-```
+## Specification
 
-### **CustomFixedFee**
+### Overview
 
-The `CustomFixedFee` message defines the type of fee and the account receiving the fee assessed during a message submission to the associated topic. A custom fee may only be fixed and must specify a fee collector account to receive the assessed fees.
+We propose adding a fixed fee mechanism to the Hedera Consensus Service (HCS) for topic messages, similar to the custom fee structures in the Hedera Token Service (HTS). This will involve creating new protocol buffers (protobuf) messages and modifying existing ones to accommodate fixed fees for topics.
 
-```
-message CustomFixedFee {
-  // Fixed fee to be charged
-  FixedFee fixed_fee = 1;
+### Requirements and Scope
 
-  // The account to receive the custom fee
-  AccountID fee_collector_account_id = 3;
-}
-```
+* Topics can have a Fee Schedule Key set at creation time.
+* Fee Schedule Key can manage fee updates and send free messages to the topic.
+* The Fee Schedule Key can be updated with an update topic transaction signed by the Admin Key and the new Fee Schedule Key.
+* If the topic was created without a Fee Schedule Key, the key cannot be added later.
+* Topics can have fees for submitting messages.
+* Fees can be set at creation time.
+* Fees can be changed (updated or removed) from a topic with an update topic transaction signed by the Fee Schedule Key.
+* Fees are defined as a list of custom fee.
+* A custom fee can be set in HBAR or HTS fungible tokens, and must have an accountID as a receiver.
+* Fee payer must set an allowance in HBAR or HTS tokens to be used to pay fees for message submissions, on an HCS topic basis.
+* Allowance granted to topic ID does not allow the topic's Admin Key (or any of the topic's keys) to spend that allowance.
+* Fee payer must set can set a maximum fee per message, on an HCS topic basis.
+* Messages signed by the Fee Schedule Key or by a Submit Key don't incur fees.
+* Applications must set the `allowCustomFeesPayment` flag to true.
+* Default value of the `allowCustomFeesPayment` flag is false.
+* If the `allowCustomFeesPayment` flag is not set, or is not true, a transaction error will be fired.
 
-### **Modified Protobuf Messages**
+### User Flows and Interaction
 
-### **TopicCreateTransactionBody**
+* Users will specify the fee settings during the topic creation process through a simple interface in their Hedera client (refer to the creation of token custom fees/fixed fee for reference).
+* Before submitting a message to a topic through an application or wallet interface, users must set an allowance and a maximum fee per message.
```

Comment:
I'm not sure if allowance should be required. I think per message fee makes sense, but in most cases where a user is sending 1 message through a wallet, requiring a max allowance, and a per message fee seems like high friction. 

I like having alowance to cap teh total amount for backend systems, but I'm not sure it should be required to be set to send messages to the topic.

Reply:
The idea is, because you need to send a tx to set a max per message in any case, you can send at the same time a max overall, and the UX does not change (you don't need to send an additional transaction). If you don't care about a max overall, you can just put a very high max, but if you care, you can set it and leverage also that limit.